<a href="https://colab.research.google.com/github/Rashmi706/Intro-to-ML-Rashmi-K-S/blob/master/Regression_redwinequality.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Introduction**

  This is a project where we need to compare several Regression algorithms to predict wine quality which has score between 0 and 10.We need to compare and select an algorithm to find out what makes a good wine by using winequality-red.csv.

**Contents of data file**

The attributes in this dataset are
  1.  fixed acidity
  2.  volatile acidity
  3.  citric acid
  4.  residual sugar
  5.  chlorides
  6.  free sulfur dioxide
  7.  total sulfur dioxide
  8.  density
  9.  pH
  10. sulphates
  11. alcohol
  12. quality(score between 0 and 10)


  **Data preprocessing**

  In this step of analysis there are some of the features to train and test the machine learning model and the target to predict which is 'quality' that are defined.Preprocessing is done using the polynomial features where these features may impact the performance of the machine learning models.Preprocessing includes finding the null values if present,checking datatypes,encoding mechanisms.


In [ ]:
import pandas as pd
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
data = pd.read_csv('/content/winequality-red.csv')

In [ ]:
data

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0               7.4             0.700         0.00  ...       0.56      9.4        5
1               7.8             0.880         0.00  ...       0.68      9.8        5
2               7.8             0.760         0.04  ...       0.65      9.8        5
3              11.2             0.280         0.56  ...       0.58      9.8        6
4               7.4             0.700         0.00  ...       0.56      9.4        5
...             ...               ...          ...  ...        ...      ...      ...
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[1599 rows x 12 columns]

In [ ]:
data.describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

In [ ]:
data.head()

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [ ]:
X = dataset.iloc[: , :-1].values
y = dataset.iloc[: , -1].values

In [ ]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [ ]:
y

array([5, 5, 5, ..., 6, 5, 6])

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

In [ ]:
X_train

array([[ 8.8 ,  0.46,  0.45, ...,  3.32,  0.79, 14.  ],
       [ 6.9 ,  0.39,  0.24, ...,  3.44,  0.58, 11.4 ],
       [ 7.4 ,  0.55,  0.19, ...,  3.49,  0.68, 10.5 ],
       ...,
       [ 9.6 ,  0.56,  0.23, ...,  3.3 ,  0.65, 10.1 ],
       [10.5 ,  0.42,  0.66, ...,  3.24,  0.75, 11.7 ],
       [ 6.1 ,  0.64,  0.02, ...,  3.47,  0.45, 11.  ]])

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR

In [ ]:
m_reg = LinearRegression()
p_reg = LinearRegression()
d_reg = DecisionTreeRegressor()
r_reg = RandomForestRegressor(n_estimators=500)
s_reg =SVR()

In [ ]:
X_poly = PolynomialFeatures(degree =2)
X_poly = X_poly.fit_transform(X_train)

In [ ]:
m_reg.fit(X_train, y_train)
p_reg.fit(X_poly,y_train)
d_reg.fit(X_train, y_train)
r_reg.fit(X_train, y_train)
s_reg.fit(X_train, y_train)

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [ ]:
temp = PolynomialFeatures(degree =2)
temp = temp.fit_transform(X_test)

In [ ]:
m_pred = m_reg.predict(X_test)
p_pred = p_reg.predict(temp)
d_pred = d_reg.predict(X_test)
r_pred = r_reg.predict(X_test)
s_pred = s_reg.predict(X_test)

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
m = r2_score(y_test, m_pred)
p = r2_score(y_test, p_pred)
d = r2_score(y_test, d_pred)
r = r2_score(y_test, r_pred)
s = r2_score(y_test, s_pred)

In [ ]:
print('Logistic Regression: ' + str(m) + '\npolyfeat: ' + str(p) + '\nDecision Tree: ' + str(d) + '\nRandom Forest: ' + str(r) + '\nSVR: ' + str(s))

Logistic Regression: 0.3785931092981135
polyfeat: 0.40628204370693677
Decision Tree: 0.15999028890507416
Random Forest: 0.5502124399126
SVR: 0.17456174576415207


**Algorithms tried on data**
1. Logistic Regression
2. Polynomial Features Logistic Regression
3. Decision Tree Regression
4. Random Forest Regression
5. Support Vector Regression

**Results**

According to the given dataset,the following results are obtained

   Logistic Regression: 0.3785931092981135

   polyfeat: 0.40628204370693677

   Decision Tree: 0.15999028890507416

   Random Forest: 0.5502124399126

   SVR: 0.17456174576415207

**Conclusion**

Based on the results we come to a conclusion that not all input features are essential and effect the data.For finding out the wine quality, we have implemented multiple algorithms.We were able to achieve maximum accuracy using Random Forest Regression algorithm.